Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/CosmosServerlessSparkBatchJob.java

```diff
@@ -143,58 +160,191 @@ public String getJobUuid() {
                 });
     }
 
-    @Override
-    public Observable<SparkBatchJobResponsePayload> getStatus() {
+    private Observable<com.microsoft.azure.hdinsight.sdk.rest.azure.serverless.spark.models.SparkBatchJob> getSparkBatchJob() {
         return getAccount().getSparkBatchJobRequest(getJobUuid())
                 .flatMap(sparkBatchJob -> {
-                    if (sparkBatchJob.properties() == null) {
-                        return Observable.error(new IOException("Can't get job status from empty property body."));
+                    if (sparkBatchJob.schedulerState() != null &&
+                            (sparkBatchJob.schedulerState().equals(SchedulerState.FINALIZING) ||
+                                    sparkBatchJob.schedulerState().equals(SchedulerState.ENDED))) {
+                        String errorMsg = "Job is in " + sparkBatchJob.schedulerState().toString() + " state.";
+                        log().warn(errorMsg);
+                        return Observable.error(new SparkJobFinishedException(errorMsg));
                     } else {
-                        return Observable.just(sparkBatchJob.properties().responsePayload());
+                        return Observable.just(sparkBatchJob);
                     }
                 });
     }
 
-    @Nullable
-    @Override
-    public String getState() throws IOException {
-        return getStatus()
-                .retry(getRetriesMax())
-                .repeatWhen(ob -> ob.delay(getDelaySeconds(), TimeUnit.SECONDS))
-                .map(sparkSubmitResponse -> sparkSubmitResponse.getState())
-                .toBlocking()
-                .singleOrDefault("unknown");
+
+    @NotNull
+    private Observable<SparkBatchJobResponsePayload> getResponsePayloadWithState() {
+        return getSparkBatchJob()
+                .filter(sparkBatchJob -> sparkBatchJob.properties() != null
```

Comment:
Also move out the filter

Reply:
If we move out the filter, the following `.map(sparkBatchJob -> sparkBatchJob.properties().responsePayload());` will encounter NPE error.